base/reduce.jl

```diff
@@ -296,46 +296,95 @@ function extrema(itr)
     return (vmin, vmax)
 end
 
-## all & any
+## any
 
-function mapfoldl(f, ::AndFun, itr)
+any(itr)    = any(IdFun(), itr)
+any(f, itr) = any(specialized_unary(f), itr)
+
+function any(f::IdFun, itr)
     for x in itr
-        !f(x) && return false
+        x && return true
     end
-    return true
+    return false
 end
 
-function mapfoldl(f, ::OrFun, itr)
+function any(f::Func{1}, itr)
     for x in itr
         f(x) && return true
     end
     return false
 end
 
-function mapreduce_impl(f, op::AndFun, A::AbstractArray{Bool}, ifirst::Int, ilast::Int)
-    while ifirst <= ilast
-        @inbounds x = A[ifirst]
-        !f(x) && return false
-        ifirst += 1
+function any(f::IdFun, itr::Array{Bool, 1})
+    i   = 1
+    len = length(itr)
+    while i <= len
+       @inbounds x = itr[i]
+       x && return true
+       i += 1
     end
-    return true
+    return false
 end
 
-function mapreduce_impl(f, op::OrFun, A::AbstractArray{Bool}, ifirst::Int, ilast::Int)
-    while ifirst <= ilast
-        @inbounds x = A[ifirst]
-        f(x) && return true
-        ifirst += 1
+function any(f::IdFun, itr::Array)
+    error("any(itr) expects the elements of itr to be booleans.")
+end
+
+function any{T}(f::Func{1}, itr::Array{T})
+    i   = 1
+    len = length(itr)
+    while i <= len
+       @inbounds x = itr[i]
+       f(x) && return true
+       i += 1
     end
     return false
 end
 
-all(a) = mapreduce(IdFun(), AndFun(), a)
-any(a) = mapreduce(IdFun(), OrFun(), a)
+## all
+
+all(itr)    = all(IdFun(), itr)
+all(f, itr) = all(specialized_unary(f), itr)
+
+function all(f::IdFun, itr)
+    for x in itr
+        !x && return false
+    end
+    return true
+end
+
+function all(f::Func{1}, itr)
+    for x in itr
+        !f(x) && return false
+    end
+    return true
+end
 
-all(pred::Union{Callable,Func{1}}, a) = mapreduce(pred, AndFun(), a)
-any(pred::Union{Callable,Func{1}}, a) = mapreduce(pred, OrFun(), a)
+function all(f::IdFun, itr::Array{Bool})
+    i   = 1
+    len = length(itr)
+    while i <= len
+       @inbounds x = itr[i]
+       !x && return false
+       i += 1
+    end
+    return true
+end
+
+function all(f::IdFun, itr::Array)
+    error("all(itr) expects the elements of itr to be booleans.")
```

Comment:
@ScottPJones  Oh no, BitArrays should stay in my opinion, if you look at the performance tests, they are the fastest of them all :P And they have been always short-circuiting, so an example to follow, not to reprimand.

Since any used to just be pretty much a alias for `reduce(|, itr)`, they both short-circuited on the same cases. It won't anymore, to let people depend on the property of reduce to always go through all elements. Not a dumb question, it's pretty confusing, which is what we're trying to fix here :)

@hayd I think so too, and `any/all` returning numbers always seemed weird to me. Should I will go ask for opnions on the issue?